src/java/org/apache/cassandra/transport/Message.java

```diff
@@ -214,9 +215,14 @@ protected boolean isTraceable()
             return false;
         }
 
-        protected abstract Response execute(QueryState queryState, long queryStartNanoTime, boolean traceRequest);
+        protected CompletableFuture<Response> executeAsync(QueryState queryState, long queryStartNanoTime, boolean traceRequest)
```

Comment:
ideally, both async and sync version would be abstract, so that the subclasses are explicit on their behaviour.
but I agree it's a good compromise to have a default implementation of `executeAsync` in the code we have.

However, please add a javadoc explaining the new flow. My understanding is: 
From the `Dispatcher` we now call `maybeExecuteAsync`, which returns a `CompletableFuture`. It is the actual implementation of a message handler that decides if request handling is blocking (the returned future is completed) or not (the handler gets scheduled and the future can be completed at a later time). The default implementation is sync execution, but for such and such queries we do an async execution.

Can we make `executeSync` non public? IIUC the dispatcher is not really supposed to call it (it doesn't in init, but does it have to?)

Reply:
... and if you want to be more fancy we could have two abstract classes - one for sync and one for async processing, with the default implementations of sync and async, respectively ;)

Reply:
> ideally, both async and sync version would be abstract, so that the subclasses are explicit on their behaviour.
but I agree it's a good compromise to have a default implementation of executeAsync in the code we have.

Agreed, I just tried to be pragmatic and strike a good tradeoff to limit changes to existing code.

> However, please add a javadoc explaining the new flow. My understanding is:

I definitely will once we agree to go down this path. And yes, your understanding is correct.

> Can we make executeSync non public?

I think it makes sense to have it public as an alternative to `maybeExecuteAsync`, if supported; speaking of which, you made me realize `executeSync` implementations throwing `ISE` should refer to `maybeExecuteAsync` because `executeAsync` is not public, I'll fix that post testing. Re: the init handler, it doesn't _have to_, but we know the startup message is sync so it made sense to me and reduced the amount of changes, I can change that though if you insist.

Reply:
> and if you want to be more fancy we could have two abstract classes - one for sync and one for async processing, with the default implementations of sync and async, respectively

C* is not the codebase to be fancy, in particular when you're dealing with existing stuff :)

Reply:
> I think it makes sense to have it public as an alternative to maybeExecuteAsync

Why should I have an alternative? Why should I learn to choose which one is better suited for my needs? I'm lazy - I don't want to learn which one is better if I can help it. `maybeExecuteAsync` makes the decision for me, and the price is having to deal with a future.

If you say "the sync version is convenient when you don't want to deal with the futures, and you know your request is handled sync" then OK, I accept it.